src/ir/match.h

```diff
@@ -0,0 +1,782 @@
+  using matched_t = Unary*;
+  using data_t = typename T::Op;
+};
+template<class T> struct MatchSelf<UnaryKind<T>> {
+  bool operator()(Unary* curr, typename T::Op op) {
+    return curr->op == T::getOp(curr->value->type, op);
+  }
+};
+template<class T> struct NumComponents<UnaryKind<T>> {
+  static constexpr size_t value = 1;
+};
+template<class T> struct GetComponent<UnaryKind<T>, 0> {
+  Expression*& operator()(Unary* curr) { return curr->value; }
+};
+template<class S> decltype(auto) UnaryMatcher(Unary** binder, UnaryOp op, S s) {
+  return Matcher<UnaryKind<UnaryK>, S>(binder, op, s);
+}
+template<class S>
+decltype(auto) AbstractUnaryMatcher(Unary** binder, Abstract::Op op, S s) {
+  return Matcher<UnaryKind<AbstractUnaryK>, S>(binder, op, s);
+}
+
+// Binary and AbstractBinary
+struct BinaryK {
+  using Op = BinaryOp;
+  static BinaryOp getOp(Type, Op op) { return op; }
+};
+struct AbstractBinaryK {
+  using Op = Abstract::Op;
+  static BinaryOp getOp(Type type, Abstract::Op op) {
+    return Abstract::getBinary(type, op);
+  }
+};
+template<class T> struct BinaryKind {};
+template<class T> struct KindTypeRegistry<BinaryKind<T>> {
+  using matched_t = Binary*;
+  using data_t = typename T::Op;
+};
+template<class T> struct MatchSelf<BinaryKind<T>> {
+  bool operator()(Binary* curr, typename T::Op op) {
+    return curr->op == T::getOp(curr->left->type, op);
+  }
+};
+template<class T> struct NumComponents<BinaryKind<T>> {
+  static constexpr size_t value = 2;
+};
+template<class T> struct GetComponent<BinaryKind<T>, 0> {
+  Expression*& operator()(Binary* curr) { return curr->left; }
+};
+template<class T> struct GetComponent<BinaryKind<T>, 1> {
+  Expression*& operator()(Binary* curr) { return curr->right; }
+};
+template<class S1, class S2>
+decltype(auto) BinaryMatcher(Binary** binder, BinaryOp op, S1 s1, S2 s2) {
+  return Matcher<BinaryKind<BinaryK>, S1, S2>(binder, op, s1, s2);
+}
+template<class S1, class S2>
+decltype(auto)
+AbstractBinaryMatcher(Binary** binder, Abstract::Op op, S1 s1, S2 s2) {
+  return Matcher<BinaryKind<AbstractBinaryK>, S1, S2>(binder, op, s1, s2);
+}
+
+// Select
+template<> struct NumComponents<Select*> { static constexpr size_t value = 3; };
+template<> struct GetComponent<Select*, 0> {
+  Expression*& operator()(Select* curr) { return curr->ifTrue; }
+};
+template<> struct GetComponent<Select*, 1> {
+  Expression*& operator()(Select* curr) { return curr->ifFalse; }
+};
+template<> struct GetComponent<Select*, 2> {
+  Expression*& operator()(Select* curr) { return curr->condition; }
+};
+template<class S1, class S2, class S3>
+decltype(auto) SelectMatcher(Select** binder, S1 s1, S2 s2, S3 s3) {
+  return Matcher<Select*, S1, S2, S3>(binder, {}, s1, s2, s3);
+}
+
+} // namespace Internal
+
+// Public matching API
+
+decltype(auto) i32() {
+  return Internal::ConstMatcher(
+    nullptr, Internal::I32Lit(nullptr, Internal::Any<int32_t>(nullptr)));
+}
+// Use int rather than int32_t to disambiguate literal 0
```

Comment:
Since `0` can be implicitly converted to any integral type or any pointer type pointer type, there is an overload resolution ambiguity when there is both an overload taking a pointer type and an overload taking a integral type. This ambiguity is solved by adding an overload for `int`, which is the original type of the literal `0`.

I will elaborate in the comment.